File: backend/app/services/scenario_service.py
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.models.scenario import Scenario, ScenarioRun


async def get_scenarios(db: AsyncSession) -> list[Scenario]:
    result = await db.execute(
        select(Scenario).order_by(Scenario.updated_at.desc())
    )
    return list(result.scalars().all())


async def get_scenario(db: AsyncSession, scenario_id: str) -> Scenario | None:
    result = await db.execute(
        select(Scenario).where(Scenario.id == scenario_id)
    )
    return result.scalar_one_or_none()


async def create_scenario(
    db: AsyncSession, name: str, description: str | None = None, graph_data: dict | None = None
) -> Scenario:
    scenario = Scenario(
        name=name,
        description=description,
        graph_data=graph_data or {},
    )
    db.add(scenario)
    await db.commit()
    await db.refresh(scenario)
    return scenario


async def update_scenario(
    db: AsyncSession,
    scenario_id: str,
    name: str | None = None,
    description: str | None = None,
    graph_data: dict | None = None,
) -> Scenario | None:
    scenario = await get_scenario(db, scenario_id)
    if not scenario:
        return None
    if name is not None:
        scenario.name = name
    if description is not None:
        scenario.description = description
    if graph_data is not None:
        scenario.graph_data = graph_data
    await db.commit()
    await db.refresh(scenario)
    return scenario


async def duplicate_scenario(db: AsyncSession, scenario_id: str) -> Scenario | None:
    original = await get_scenario(db, scenario_id)
    if not original:
        return None
    copy = Scenario(
        name=f"{original.name} (копия)",
        description=original.description,
        graph_data=original.graph_data,
    )
    db.add(copy)
    await db.commit()
    await db.refresh(copy)
    return copy


async def delete_scenario(db: AsyncSession, scenario_id: str) -> bool:
    scenario = await get_scenario(db, scenario_id)
    if not scenario:
        return False
    await db.delete(scenario)
    await db.commit()
    return True


async def get_scenario_runs(db: AsyncSession, scenario_id: str) -> list[ScenarioRun]:
    result = await db.execute(
        select(ScenarioRun)
        .where(ScenarioRun.scenario_id == scenario_id)
        .order_by(ScenarioRun.started_at.desc())
    )
    return list(result.scalars().all())


async def get_scenario_run(db: AsyncSession, run_id: str) -> ScenarioRun | None:
    result = await db.execute(
        select(ScenarioRun)
        .where(ScenarioRun.id == run_id)
        .options(selectinload(ScenarioRun.steps))
    )
    return result.scalar_one_or_none()

File: backend-java/src/main/java/com/gigame/service/ScenarioService.java
package com.gigame.service;

import com.gigame.dto.scenario.*;
import com.gigame.model.Scenario;
import com.gigame.model.ScenarioRun;
import com.gigame.model.ScenarioRunStep;
import com.gigame.repository.ScenarioRepository;
import com.gigame.repository.ScenarioRunRepository;
import com.gigame.repository.ScenarioRunStepRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Simple CRUD operations for scenarios. No execution logic here.
 */
@Service
public class ScenarioService {

    private final ScenarioRepository scenarioRepository;
    private final ScenarioRunRepository scenarioRunRepository;
    private final ScenarioRunStepRepository scenarioRunStepRepository;

    public ScenarioService(ScenarioRepository scenarioRepository,
                           ScenarioRunRepository scenarioRunRepository,
                           ScenarioRunStepRepository scenarioRunStepRepository) {
        this.scenarioRepository = scenarioRepository;
        this.scenarioRunRepository = scenarioRunRepository;
        this.scenarioRunStepRepository = scenarioRunStepRepository;
    }

    // -------------------------------------------------------------------------
    // Scenario CRUD
    // -------------------------------------------------------------------------

    public List<ScenarioResponse> getScenarios() {
        return scenarioRepository.findAllByOrderByUpdatedAtDesc().stream()
                .map(this::toResponse)
                .toList();
    }

    public ScenarioDetailResponse getScenario(UUID id) {
        Scenario scenario = scenarioRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Scenario not found: " + id));
        return toDetailResponse(scenario);
    }

    @Transactional
    public ScenarioResponse createScenario(String name, String description, Map<String, Object> graphData) {
        Scenario scenario = new Scenario();
        scenario.setName(name);
        scenario.setDescription(description);
        scenario.setGraphData(graphData != null ? graphData : Map.of());
        scenario = scenarioRepository.save(scenario);
        return toResponse(scenario);
    }

    @Transactional
    public ScenarioResponse updateScenario(UUID id, String name, String description, Map<String, Object> graphData) {
        Scenario scenario = scenarioRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Scenario not found: " + id));
        if (name != null) {
            scenario.setName(name);
        }
        if (description != null) {
            scenario.setDescription(description);
        }
        if (graphData != null) {
            scenario.setGraphData(graphData);
        }
        scenario = scenarioRepository.save(scenario);
        return toResponse(scenario);
    }

    @Transactional
    public ScenarioResponse duplicateScenario(UUID id) {
        Scenario original = scenarioRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Scenario not found: " + id));
        Scenario copy = new Scenario();
        copy.setName(original.getName() + " (\u043a\u043e\u043f\u0438\u044f)");
        copy.setDescription(original.getDescription());
        copy.setGraphData(original.getGraphData());
        copy = scenarioRepository.save(copy);
        return toResponse(copy);
    }

    @Transactional
    public void deleteScenario(UUID id) {
        Scenario scenario = scenarioRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Scenario not found: " + id));
        scenarioRepository.delete(scenario);
    }

    // -------------------------------------------------------------------------
    // Scenario Run queries
    // -------------------------------------------------------------------------

    public List<ScenarioRunResponse> getScenarioRuns(UUID scenarioId) {
        return scenarioRunRepository.findByScenarioIdOrderByStartedAtDesc(scenarioId).stream()
                .map(this::toRunResponse)
                .toList();
    }

    public ScenarioRunDetailResponse getScenarioRun(UUID runId) {
        ScenarioRun run = scenarioRunRepository.findById(runId)
                .orElseThrow(() -> new IllegalArgumentException("Scenario run not found: " + runId));
        List<ScenarioRunStepResponse> stepResponses = scenarioRunStepRepository
                .findByRunIdOrderByStartedAtAsc(run.getId()).stream()
                .map(this::toStepResponse)
                .toList();
        return new ScenarioRunDetailResponse(
                run.getId(),
                run.getScenarioId(),
                run.getStatus(),
                run.getInputDocumentIds(),
                run.getResult(),
                run.getStartedAt(),
                run.getCompletedAt(),
                stepResponses
        );
    }

    // -------------------------------------------------------------------------
    // Mapping helpers
    // -------------------------------------------------------------------------

    private ScenarioResponse toResponse(Scenario s) {
        return new ScenarioResponse(
                s.getId(),
                s.getName(),
                s.getDescription(),
                s.getCreatedAt(),
                s.getUpdatedAt()
        );
    }

    private ScenarioDetailResponse toDetailResponse(Scenario s) {
        return new ScenarioDetailResponse(
                s.getId(),
                s.getName(),
                s.getDescription(),
                s.getCreatedAt(),
                s.getUpdatedAt(),
                s.getGraphData()
        );
    }

    private ScenarioRunResponse toRunResponse(ScenarioRun r) {
        return new ScenarioRunResponse(
                r.getId(),
                r.getScenarioId(),
                r.getStatus(),
                r.getInputDocumentIds(),
                r.getResult(),
                r.getStartedAt(),
                r.getCompletedAt()
        );
    }

    private ScenarioRunStepResponse toStepResponse(ScenarioRunStep s) {
        return new ScenarioRunStepResponse(
                s.getId(),
                s.getNodeId(),
                s.getNodeType(),
                s.getStatus(),
                s.getInputData(),
                s.getOutputData(),
                s.getPromptUsed(),
                s.getTokensUsed(),
                s.getStartedAt(),
                s.getCompletedAt()
        );
    }
}
